src/scheduler/MpiWorld.cpp

```diff
@@ -714,17 +787,65 @@ void MpiWorld::awaitAsyncRequest(int requestId)
 {
     SPDLOG_TRACE("MPI - await {}", requestId);
 
-    auto it = futureMap.find(requestId);
-    if (it == futureMap.end()) {
-        throw std::runtime_error(
-          fmt::format("Error: waiting for unrecognized request {}", requestId));
+    // Get the corresponding send and recv ranks
+    auto it = reqIdToRanks.find(requestId);
+    if (it == reqIdToRanks.end()) {
+        SPDLOG_ERROR("Asynchronous request id not recognized: {}", requestId);
+        throw std::runtime_error("Unrecognized async request id");
+    }
+    int sendRank = it->second.first;
+    int recvRank = it->second.second;
+    reqIdToRanks.erase(it);
+
+    // If awaiting an isend request, return immediately as our transport layer
+    // is asynchronous, thus request is already handled
+    if (sendRank == MPI_IS_ISEND_REQUEST && recvRank == MPI_IS_ISEND_REQUEST) {
```

Comment:
Putting this `MPI_IS_ISEND_REQUEST` constant in as a marker for isend requests is a bit hard to read; it's overloading the `reqIdToRanks` map with two jobs. Instead it might be easier to understand and reason about it we had a separate `std::set` to track the IDs of isend requests, then this method would start with:

```
if(iSendRequests.find(requestId) == iSendRequests.end()) {
    return;
}
```

I also think debugging might get confusing as the value of the constant is `-1`, so you might have to take a few steps to realise what's going on to get a rank of `-1`.